### ISSUE TITLE
Both load_ini_path and load_ini_file method calls fail from C++ code

### ISSUE THREAD

**Issue by jjochum-ELT** [madmurphy/libconfini#6]
I have built **libconfini** and referenced the generated file `libconfini\src\.libs\libconfini-0.dll` in an empty **C++ Managed Build** project using **Eclipse CDT** on a **Windows 7** host with **mingw64** (gcc version 8.2).

Building my project _TestIniReader_ works fine, but when calling my application the output is always:

```
The file content is:
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
Sorry, something went wrong :-(
To quit, press 'q' then the 'Enter' key.
```

The source code (see [TestIniReader.zip](https://github.com/madmurphy/libconfini/files/2465013/TestIniReader.zip)) is pretty much the same as mentioned in the documentation of **libconfini**, still the `callback` method does not seem to get called. The line "This is a message!" is never printed to console.

How can I find out what the couse for the failed method call to `load_ini_path` is?

**Comment by madmurphy** [comment]
@jjochum-ELT

Thanks. I am not at home at the moment, I will try to have a look at your code later. In the meanwhile:

1. What is the error code returned by `load_ini_path()`?
2. Are you sure that the file is not still used by the lines of code you have added for printing the file content? Does it still not work if you try not to access the file before calling `load_ini_path()`?

--madmurphy

**Comment by madmurphy** [comment]
@jjochum-ELT

I have tried your code on linux and it worked. This makes me believe the problem *is* actually that you don't close the stream before trying to access the same file (Windows is way more whimsical with that). Can you try to close it before running `load_ini_path()`?
```c
	// ...

	f.close();

	// Read and print contents from INI file.

	// ...
```

**Comment by jjochum-ELT** [comment]
Right, I forgot to close the stream, but closing it does not help :(. I actually just added the `// Check if config.ini file exists and if so print the content.` part after I had no success in successfully running the `// Read and print contents from INI file.` segment. 

Here is an excerpt of my updated main():
```
f.close();

// Read and print contents from INI file.
if (int res = load_ini_path("config.ini", INI_DEFAULT_FORMAT, NULL,
		callback,
		NULL)) {

	fprintf(stderr, "Sorry, something went wrong :-( - ERROR %i\n", res);
}
```
Interestingly I get a different error code when running the code using **Eclipse CDT**'s Debug feature, to when I start the compiled _TestIniReader.exe_ using the Windows Explorer.

The output printed on the CMD directly starting the _TestIniReader.exe_ from the Windows Explorer is:
```
The file content is:
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
Sorry, something went wrong :-( - ERROR 6
To quit, press 'q' then the 'Enter' key.
```

The output printed on the CMD when using Eclipse CDT is:
```
Sorry, something went wrong :-( - ERROR 4
To quit, press 'q' then the 'Enter' key.
```
So the error code is different and in the second case the whole `The file content is:` output is missing for some reason. But that is probably unrelated to the INI file not being read by your library, but it might have something to do with the difference in the error codes. I will have to figure out whats going on there. 

Pretty frustrating that something that simple does not immediately work... :(

**Comment by madmurphy** [comment]
@jjochum-ELT

These two error codes mean "File inaccessible" and "Error reading the file" (you can have a look at the `enum` `ConfiniInterruptNo` in **confini.h**). They are returned by `fopen()` and `fread()` *before* the parsing begins (these two functions are invoked immediately at the beginning of `load_ini_path()` and `load_ini_file()`).

The only thing I can say for now is that this problem has nothing to do with the parsing, but with file access instead, and the solution might be quite trivial. I will try to reproduce your problem in the next days using Windows.

In the meanwyhile you can try to understand what is going on at the beginning of `load_ini_path()` and `load_ini_file()` (by searching for `CONFINI_EIO` and `CONFINI_ENOENT`).

--madmurphy

**Comment by madmurphy** [comment]
@jjochum-ELT

I got a suspect. Could you recompile the library after replacing current line 2151

```c
	FILE * const ini_file = fopen(path, "r");
```

with this new version of the same line?

```c
	FILE * const ini_file = fopen(path, "rb");
```

--madmurphy

**Comment by jjochum-ELT** [comment]
Ok, I found out the problem was the **Windows** style line endings (CRLF) in the _config.ini_ file.

I updated my code, which is now using the `load_ini_file` method and also reading and printing the file contents using C methods.
Additionally I added two _config\_*.ini_ files, one with **Windows** style line endings (CRLF) and one with **Linux** style line endings (LF). The source code and config files are appended here: [TestIniReader.zip](https://github.com/madmurphy/libconfini/files/2469241/TestIniReader.zip).

### Running in Eclipse CDT

When setting either filename `string fileName = "config_win.ini";` or `string fileName = "config_linux.ini";` the output in the console opened by **Eclipse CDT** shows:
```
File errorfile length is -1
Memory errorresult is 0
Reading error
ERROR result:0 not equal to lSize:-1
Printing file content using C methods:
```
So the file length of -1 is somehow completely messed up, which makes running code in Eclipse CDT unusable.

### Running by starting TestIniReader.exe from Windows Explorer
When setting filename `string fileName = "config_win.ini";` the output in the CMD console shows:

```
file length is 83
result is 79
Reading error
ERROR result:79 not equal to lSize:83
Printing file content using C methods:
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
The file content is:
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
test 6
Sorry, something went wrong :-( - ERROR 6
To quit, press 'q' then the 'Enter' key.
```
So the length `lSize` of 83 is actually correct, but the `result` of 79 returned by the `fread` C method is wrong. For some reason it cannot deal with the 4 additional CR characters!?

I don't know whether you can find a C method compatible with CRLF line endings or whether you want to demand the user to provide a linux style file? But that might be something you could look into...

When setting filename `string fileName = "config_linux.ini";` the output in the CMD console shows:

```
file length is 79
result is 79
Printing file content using C methods:
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
The file content is:
[users]
have_visited = ronnie, lilly82, robrob

[last_update]
date = 12.03.2017
This is a message!
This is a message!
People who have visited: ronnie, lilly82, robrob
This is a message!
This is a message!
Last update: 12.03.2017
To quit, press 'q' then the 'Enter' key.
```
So this is completely fine and as expected. 

I guess it is easier for me to look for a different library (written in C++) to read INI files, since I am using a Windows system and cannot have something like that potentially break my program. Also I will switch to **Visual Studio**, **Eclipse CDT** seems to be too messed up for me.
Still I appretiate the work you have put into this and also for helping me solve my problem.

---------------------
Thanks, `fopen(path, "rb");` works with the **Windows** style line endings file

**Comment by madmurphy** [comment]
@jjochum-ELT 

Thank you, it was a very helpful discussion. I rarely use Windows, so this bug went out of my sight. I will now update the code and replace it with `fopen(path, "rb");`.

> I guess it is easier for me to look for a different library (written in C++) to read INI files, since I am using a Windows system and cannot have something like that potentially break my program.

With my fix the problem is completely solved once and for all. But you can choose something else, of course :-P

**Comment by jjochum-ELT** [comment]
Hehe :), yes you are right. Actually it should be solved.
I was just in the middle of writing my text when you proposed the solution and I only added the last line in my previous post without looking over the rest of what I have written, which has partially become obsolete. Among others:

> I don't know whether you can find a C method compatible with CRLF line endings or whether you want to demand the user to provide a linux style file? But that might be something you could look into...

**Comment by madmurphy** [comment]
:)

Consider that the IO functions are the only parts of the code that interface with the operating system, and in **libconfini** those consist of only three lines of code if I don't remember wrong. The rest of the code is just bare metal for parsing strings, so if it works in one machine it will definitely work everywhere else.

My suggestion is: just keep playing a bit with libconfini before taking any decision, you might discover it can do a lot of smart things.

**Comment by madmurphy** [comment]
:)

Consider that the IO functions are the only parts of the code that interfaces with the operating system, and in **libconfini** these consist of only three lines of code if I don't remember wrong. The rest of the code is just bare metal for parsing strings, so if it works in one machine it will definitely work everywhere else.

My suggestion is: just keep playing a bit with libconfini before taking any decision, you might discover it can do a lot of smart things.
